<a href="https://colab.research.google.com/github/jkranyak/project_3/blob/main/Cancer_Net_SCal_Project3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
!pip install tensorflow
!pip install isic-cli
!pip install kaggle
# !pip install imblearn

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 267.1/267.1 kB 12.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 223.7/223.7 kB 31.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.5/49.5 kB 7.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.5/54.5 kB 3.9 MB/s eta 0:00:00


In [3]:
import random
import numpy as np
import tensorflow as tf

# Set random seeds for Python's built-in random number generator
random.seed(42)

# Set random seeds for NumPy's random number generator
np.random.seed(42)

# Set random seeds for TensorFlow's random number generator
tf.random.set_seed(42)


In [17]:
import os
# Specify the current path of kaggle.json file
current_path = '/content/kaggle.json'

# Desired path where the Kaggle API expects the kaggle.json file
desired_path = '/root/.kaggle/kaggle.json'

if os.path.exists(current_path):
    os.makedirs(os.path.dirname(desired_path), exist_ok=True)
    os.rename(current_path, desired_path)

    # Set the file's permissions to avoid a permissions error
    os.chmod(desired_path, 0o600)
else:
    print(f"Error: '{current_path}' does not exist. Please upload the file.")



In [16]:
! isic user login

Hello jessekranyak@gmail.com!


In [18]:
!kaggle datasets download -d andrewmvd/isic-2019


100% 9.08G/9.10G [01:27<00:00, 111MB/s]
100% 9.10G/9.10G [01:27<00:00, 111MB/s]


In [41]:
!unzip -q isic-2019.zip


replace ISIC_2019_Training_GroundTruth.csv? [y]es, [n]o, [A]ll, [N]one, [r]ename: A


In [20]:
!pip install scikit-learn

In [21]:
import numpy as np
import pandas as pd
import sklearn.utils.class_weight
from pathlib import Path
from sklearn.model_selection import train_test_split
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.applications import DenseNet169
from tensorflow.keras.layers import GlobalAveragePooling2D, Dense, Dropout, BatchNormalization, Activation
from tensorflow.keras.models import Model
from tensorflow.keras.utils import to_categorical
from sklearn.utils.class_weight import compute_class_weight
from sklearn.preprocessing import LabelEncoder, OneHotEncoder



https://www.nature.com/articles/s41598-021-96707-8

get the data over here

run all the same cells and augment by resampling the data like we did in model 1


In [96]:
metadata = pd.read_csv('/content/challenge-2019-training_metadata_2024-04-17.csv')
metadata.rename(columns={'isic_id': 'image'}, inplace=True)
# Define the new base path for the images
base_path = Path('/content/ISIC_2019_Training_Input/ISIC_2019_Training_Input')

# Create the full image paths
metadata['image_path'] = metadata['image'].apply(lambda x: base_path / f"{x}.jpg")

# Example of how to display the first few image paths
print(metadata[['image', 'image_path']].head())

          image                                         image_path
0  ISIC_0000000  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...
1  ISIC_0000000  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...
2  ISIC_0000000  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...
3  ISIC_0000000  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...
4  ISIC_0000001  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...


<ipython-input-96-1c578a8e213b>:1: DtypeWarning: Columns (8,11,15,16,17) have mixed types. Specify dtype option on import or set low_memory=False.
  metadata = pd.read_csv('/content/challenge-2019-training_metadata_2024-04-17.csv')


In [97]:
metadata['image_path'].head()

0    /content/ISIC_2019_Training_Input/ISIC_2019_Tr...
1    /content/ISIC_2019_Training_Input/ISIC_2019_Tr...
2    /content/ISIC_2019_Training_Input/ISIC_2019_Tr...
3    /content/ISIC_2019_Training_Input/ISIC_2019_Tr...
4    /content/ISIC_2019_Training_Input/ISIC_2019_Tr...
Name: image_path, dtype: object

In [98]:
# Drop the specified columns
metadata.drop(columns=['copyright_license', 'attribution'], inplace=True)
# Rename the column
metadata.rename(columns={'isic_id': 'image'}, inplace=True)
metadata.drop_duplicates(inplace=True)
metadata.head()

image  age_approx anatom_site_general benign_malignant  \
0   ISIC_0000000        55.0      anterior torso           benign   
4   ISIC_0000001        30.0      anterior torso           benign   
8   ISIC_0000002        60.0     upper extremity        malignant   
12  ISIC_0000003        30.0     upper extremity           benign   
16  ISIC_0000004        80.0     posterior torso        malignant   

    clin_size_long_diam_mm  concomitant_biopsy dermoscopic_type diagnosis  \
0                      NaN               False              NaN     nevus   
4                      NaN               False              NaN     nevus   
8                      NaN                True              NaN  melanoma   
12                     NaN               False              NaN     nevus   
16                     NaN                True              NaN  melanoma   

   diagnosis_confirm_type family_hx_mm   image_type lesion_id  melanocytic  \
0                     NaN          NaN  dermoscopic       NaN         True   
4                     NaN          NaN  dermoscopic       NaN         True   
8          histopathology          NaN  dermoscopic       NaN         True   
12                    NaN          NaN  dermoscopic       NaN         True   
16         histopathology          NaN  dermoscopic       NaN         True   

   nevus_type patient_id personal_hx_mm     sex  \
0         NaN        NaN            NaN  female   
4         NaN        NaN            NaN  female   
8         NaN        NaN            NaN  female   
12        NaN        NaN            NaN    male   
16        NaN        NaN            NaN    male   

                                           image_path  
0   /content/ISIC_2019_Training_Input/ISIC_2019_Tr...  
4   /content/ISIC_2019_Training_Input/ISIC_2019_Tr...  
8   /content/ISIC_2019_Training_Input/ISIC_2019_Tr...  
12  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...  
16  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...

In [99]:
metadata.shape

(25331, 18)

In [100]:
# Load the data from the file into a DataFrame
ground_truth = pd.read_csv('/content/ISIC_2019_Training_GroundTruth.csv')

# Display the first few rows of the DataFrame
ground_truth.shape

(25331, 10)

In [101]:
image_dir = Path('/content/ISIC_2019_Training_Input/ISIC_2019_Training_Input')
ground_truth['image_path'] = ground_truth['image'].apply(lambda x: image_dir / f"{x}.jpg")

# Merge the ground_truth with metadata if necessary
full_metadata = pd.merge(ground_truth, metadata, on='image', how='left')  # Adjust 'on' parameter as needed
full_metadata.shape

(25331, 28)

In [102]:
# Correct the base path in 'image_path' column
correct_base_path = "/content/ISIC_2019_Training_Input/ISIC_2019_Training_Input"

full_metadata['image_path'] = full_metadata['image'].apply(lambda x: f"{correct_base_path}/{x}.jpg")

# Verify the correction by printing the first few entries again
print(full_metadata['image_path'].head())

0    /content/ISIC_2019_Training_Input/ISIC_2019_Tr...
1    /content/ISIC_2019_Training_Input/ISIC_2019_Tr...
2    /content/ISIC_2019_Training_Input/ISIC_2019_Tr...
3    /content/ISIC_2019_Training_Input/ISIC_2019_Tr...
4    /content/ISIC_2019_Training_Input/ISIC_2019_Tr...
Name: image_path, dtype: object


In [103]:
full_metadata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25331 entries, 0 to 25330
Data columns (total 29 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   image                   25331 non-null  object 
 1   MEL                     25331 non-null  float64
 2   NV                      25331 non-null  float64
 3   BCC                     25331 non-null  float64
 4   AK                      25331 non-null  float64
 5   BKL                     25331 non-null  float64
 6   DF                      25331 non-null  float64
 7   VASC                    25331 non-null  float64
 8   SCC                     25331 non-null  float64
 9   UNK                     25331 non-null  float64
 10  image_path_x            25331 non-null  object 
 11  age_approx              22989 non-null  float64
 12  anatom_site_general     21374 non-null  object 
 13  benign_malignant        15710 non-null  object 
 14  clin_size_long_diam_mm  556 non-null  

In [104]:
# Fill missing values for 'age_approx' with its median
full_metadata['age_approx'].fillna(full_metadata['age_approx'].median(), inplace=True)

# For categorical data, fill missing values with 'unknown'
full_metadata['anatom_site_general'].fillna('unknown', inplace=True)
full_metadata['sex'].fillna('unknown', inplace=True)
full_metadata.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25331 entries, 0 to 25330
Data columns (total 29 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   image                   25331 non-null  object 
 1   MEL                     25331 non-null  float64
 2   NV                      25331 non-null  float64
 3   BCC                     25331 non-null  float64
 4   AK                      25331 non-null  float64
 5   BKL                     25331 non-null  float64
 6   DF                      25331 non-null  float64
 7   VASC                    25331 non-null  float64
 8   SCC                     25331 non-null  float64
 9   UNK                     25331 non-null  float64
 10  image_path_x            25331 non-null  object 
 11  age_approx              25331 non-null  float64
 12  anatom_site_general     25331 non-null  object 
 13  benign_malignant        15710 non-null  object 
 14  clin_size_long_diam_mm  556 non-null  

In [105]:
# Sample DataFrame
data = {
    'image_path_x': ['path1.jpg', 'path2.jpg', 'path3.jpg'],
    'image_path_y': ['path1.jpg', 'path2.jpg', 'path3.jpg'],
    'image_path': ['path1.jpg', 'path2.jpg', 'path4.jpg']  # Notice the discrepancy in the last row
}
df = full_metadata

# Step 1: Consolidate columns into one
df['consolidated_path'] = df.apply(lambda x: x.dropna().unique()[0] if len(x.dropna().unique()) == 1 else None, axis=1)

# Step 2: Identify discrepancies
df['is_discrepant'] = df.apply(lambda x: len(x.dropna().unique()) != 1, axis=1)

# Step 3: Count discrepancies
discrepancy_count = df['is_discrepant'].sum()

# Step 4: Report discrepancies
discrepancy_report = df[df['is_discrepant']]

print("Number of discrepancies:", discrepancy_count)
print("Discrepancy report:")
print(discrepancy_report[['image_path_x', 'image_path_y', 'image_path', 'consolidated_path']])

Number of discrepancies: 25331
Discrepancy report:
                                            image_path_x  \
0      /content/ISIC_2019_Training_Input/ISIC_2019_Tr...   
1      /content/ISIC_2019_Training_Input/ISIC_2019_Tr...   
2      /content/ISIC_2019_Training_Input/ISIC_2019_Tr...   
3      /content/ISIC_2019_Training_Input/ISIC_2019_Tr...   
4      /content/ISIC_2019_Training_Input/ISIC_2019_Tr...   
...                                                  ...   
25326  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...   
25327  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...   
25328  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...   
25329  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...   
25330  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...   

                                            image_path_y  \
0      /content/ISIC_2019_Training_Input/ISIC_2019_Tr...   
1      /content/ISIC_2019_Training_Input/ISIC_2019_Tr...   
2      /content/ISIC_2019_Training_Input/ISIC_20

In [106]:
import pandas as pd

# Define a function to convert PosixPath to string and extract the last part of the paths
def detail_discrepancies(row):
    # Convert PosixPath to string and then get the last 30 characters of each path
    x_end = str(row['image_path_x'])[-30:]
    y_end = str(row['image_path_y'])[-30:]
    path_end = str(row['image_path'])[-30:]

    if x_end != y_end or x_end != path_end:
        return f"X vs Y: {x_end} != {y_end}, X vs Path: {x_end} != {path_end}"
    return None

# Apply this function to a new column
full_metadata['discrepancy_details'] = full_metadata.apply(detail_discrepancies, axis=1)

# Filter to show only rows with discrepancies and limit the output
discrepancy_details = full_metadata[full_metadata['discrepancy_details'].notna()].head(10)

print("Detailed discrepancies (first 10 cases focusing on path endings):")
print(discrepancy_details[['image_path_x', 'image_path_y', 'image_path', 'discrepancy_details']])


Detailed discrepancies (first 10 cases focusing on path endings):
                                         image_path_x image_path_y  \
16  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...          NaN   
17  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...          NaN   
18  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...          NaN   
19  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...          NaN   
20  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...          NaN   
21  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...          NaN   
22  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...          NaN   
23  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...          NaN   
24  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...          NaN   
25  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...          NaN   

                                           image_path  \
16  /content/ISIC_2019_Training_Input/ISIC_2019_Tr...   
17  /content/ISIC_2019_Training_Input/ISIC_2019_T

In [107]:
full_metadata = pd.get_dummies(full_metadata, columns=['anatom_site_general','benign_malignant', 'sex'])


In [108]:
anatom_site_columns = [col for col in full_metadata.columns if col.startswith('anatom_site_general_')]
#benign_malignant_columns = [col for col in full_metadata.columns if col.startswith('benign_malignant_')]

print("One-hot encoded columns for 'anatom_site_general':")
full_metadata.head()


One-hot encoded columns for 'anatom_site_general':


image  MEL   NV  BCC   AK  BKL   DF  VASC  SCC  UNK  ...  \
0  ISIC_0000000  0.0  1.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  ...   
1  ISIC_0000001  0.0  1.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  ...   
2  ISIC_0000002  1.0  0.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  ...   
3  ISIC_0000003  0.0  1.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  ...   
4  ISIC_0000004  1.0  0.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0  ...   

  anatom_site_general_palms/soles  anatom_site_general_posterior torso  \
0                           False                                False   
1                           False                                False   
2                           False                                False   
3                           False                                False   
4                           False                                 True   

   anatom_site_general_unknown anatom_site_general_upper extremity  \
0                        False                               False   
1                        False                               False   
2                        False                                True   
3                        False                                True   
4                        False                               False   

  benign_malignant_benign benign_malignant_indeterminate/benign  \
0                    True                                 False   
1                    True                                 False   
2                   False                                 False   
3                    True                                 False   
4                   False                                 False   

  benign_malignant_malignant sex_female sex_male sex_unknown  
0                      False       True    False       False  
1                      False       True    False       False  
2                       True       True    False       False  
3                      False      False     True       False  
4                       True      False     True       False  

[5 rows x 44 columns]

In [109]:
# Rename the one-hot encoded 'benign_malignant' columns
full_metadata.rename(columns={
     'benign_malignant_benign': 'benign',
     'benign_malignant_indeterminate/benign': 'indeterminate_benign',
     'benign_malignant_malignant': 'malignant'
 }, inplace=True)

# Sum the 'indeterminate_benign' and 'malignant' columns to create a single 'malignant' column
full_metadata['malignant'] = full_metadata['indeterminate_benign'] + full_metadata['malignant']

# Convert any values greater than 1 back to 1 (if there's any overlapping)
full_metadata['malignant'] = full_metadata['malignant'].clip(upper=1)

# Drop the now unnecessary 'indeterminate_benign' column as it has been merged into 'malignant'
full_metadata.drop(columns='indeterminate_benign', inplace=True)

# Verify the changes
print(full_metadata[['benign', 'malignant']].head())
full_metadata.info()

   benign  malignant
0    True      False
1    True      False
2   False       True
3    True      False
4   False       True
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25331 entries, 0 to 25330
Data columns (total 43 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   image                                25331 non-null  object 
 1   MEL                                  25331 non-null  float64
 2   NV                                   25331 non-null  float64
 3   BCC                                  25331 non-null  float64
 4   AK                                   25331 non-null  float64
 5   BKL                                  25331 non-null  float64
 6   DF                                   25331 non-null  float64
 7   VASC                                 25331 non-null  float64
 8   SCC                                  25331 non-null  float64
 9   UNK                               

In [110]:
# Replace True and False with 1 and 0 respectively
full_metadata.replace({True: 1, False: 0}, inplace=True)

# Drop the specified columns
columns_to_drop = ['diagnosis', 'nevus_type','patient_id', 'dermoscopic_type', 'diagnosis_confirm_type', 'image_type', 'image_path_x', 'image_path_y', 'lesion_id']
full_metadata.drop(columns_to_drop, axis=1, inplace=True)

# Perform Label Encoding for the specified columns
label_encoder = LabelEncoder()
for column in ['concomitant_biopsy', 'family_hx_mm', 'melanocytic', 'personal_hx_mm', 'discrepancy_details', 'consolidated_path', 'clin_size_long_diam_mm' ]:
    full_metadata[column] = label_encoder.fit_transform(full_metadata[column])

full_metadata = full_metadata.set_index('image')

# Displaying the modified DataFrame
print(full_metadata.head())

              MEL   NV  BCC   AK  BKL   DF  VASC  SCC  UNK  age_approx  ...  \
image                                                                   ...   
ISIC_0000000  0.0  1.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0        55.0  ...   
ISIC_0000001  0.0  1.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0        30.0  ...   
ISIC_0000002  1.0  0.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0        60.0  ...   
ISIC_0000003  0.0  1.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0        30.0  ...   
ISIC_0000004  1.0  0.0  0.0  0.0  0.0  0.0   0.0  0.0  0.0        80.0  ...   

              anatom_site_general_oral/genital  \
image                                            
ISIC_0000000                                 0   
ISIC_0000001                                 0   
ISIC_0000002                                 0   
ISIC_0000003                                 0   
ISIC_0000004                                 0   

              anatom_site_general_palms/soles  \
image                                         

In [111]:
label_encoder = LabelEncoder()
for column in ['age_approx']:
    full_metadata[column] = label_encoder.fit_transform(full_metadata[column])


full_metadata.info()

<class 'pandas.core.frame.DataFrame'>
Index: 25331 entries, ISIC_0000000 to ISIC_0073254
Data columns (total 33 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   MEL                                  25331 non-null  float64
 1   NV                                   25331 non-null  float64
 2   BCC                                  25331 non-null  float64
 3   AK                                   25331 non-null  float64
 4   BKL                                  25331 non-null  float64
 5   DF                                   25331 non-null  float64
 6   VASC                                 25331 non-null  float64
 7   SCC                                  25331 non-null  float64
 8   UNK                                  25331 non-null  float64
 9   age_approx                           25331 non-null  int64  
 10  clin_size_long_diam_mm               25331 non-null  int64  
 11  concomitant_bio

In [112]:
!pip show Sequence

Name: sequence
Version: 0.3.4
Summary: A Python module for looping over a sequence of commands with a focus on high configurability and extensibility.
Home-page: https://github.com/thomst/sequence
Author: Thomas Leichtfuss
Author-email: thomaslfuss@gmx.de
License: GPL
Location: /usr/local/lib/python3.10/dist-packages
Requires: daytime
Required-by: 


In [113]:
!pip install Sequence

#Feature engineering

In [118]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split

# Create a copy of the data to extract the most important features
# Make sure to call the copy() method
cleaned_data = full_metadata.copy()

# Assuming cleaned_data is a DataFrame and 'malignant', 'image_path' are columns in this DataFrame
# Prepare data by dropping non-features columns
X = cleaned_data.drop(['malignant', 'image_path'], axis=1)  # Drop columns correctly in DataFrame
y = cleaned_data['malignant']  # This should be fine as it extracts a Series

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train model
rf = RandomForestClassifier(n_estimators=100, random_state=42)
rf.fit(X_train, y_train)

# Get feature importance
importances = rf.feature_importances_
feature_names = X.columns
feature_importance_dict = dict(zip(feature_names, importances))

# Print sorted feature importance
sorted_importance = sorted(feature_importance_dict.items(), key=lambda item: item[1], reverse=True)
for feature, importance in sorted_importance:
    print(f"{feature}: {importance}")


MEL: 0.5179439053785109
melanocytic: 0.13725922017686334
NV: 0.11950970770962911
benign: 0.06354317965678453
concomitant_biopsy: 0.04682150732932103
sex_unknown: 0.02583392354869595
discrepancy_details: 0.01907062605579324
anatom_site_general_unknown: 0.015879600071996482
BCC: 0.014482607144655248
BKL: 0.014183908931746203
age_approx: 0.007847932811683547
anatom_site_general_head/neck: 0.0039845219677204415
sex_male: 0.0037170403399390695
AK: 0.0023169556723333174
SCC: 0.0022226428649305373
sex_female: 0.0018210871519961107
anatom_site_general_palms/soles: 0.0009474862775256677
anatom_site_general_upper extremity: 0.0004861732895128807
DF: 0.00037590677037871234
anatom_site_general_anterior torso: 0.0003334000230439356
VASC: 0.00030277828943982396
anatom_site_general_lower extremity: 0.0002832095273032177
clin_size_long_diam_mm: 0.00027001149795925387
family_hx_mm: 0.00018478017824199194
anatom_site_general_posterior torso: 0.00017143016062070183
personal_hx_mm: 0.00016998381038237404


###Lets define the best features
Top Influential Features:

MEL (Melanoma): With an importance of approximately 51.79%, this feature dominates others by a significant margin, confirming its critical role in predicting malignancy due to its strong association with cancerous skin lesions.

melanocytic: Important at around 13.73%, indicating that features related to melanocytes are crucial for distinguishing malignant cases.

NV (Nevus): Has an importance of about 11.95%, suggesting it's also a significant predictor, likely due to its common occurrence in skin examinations but generally benign nature.

benign: This feature's importance at about 6.35% suggests it effectively helps the model distinguish non-cancerous growths, which is crucial for reducing false positives in cancer detection.

concomitant_biopsy: With an importance of 4.68%, this indicates that the presence of other biopsies performed might be indicative of more serious investigations and possibly higher chances of malignancy detection.


In [126]:
# Cancer-Net SCa Model
from sklearn.utils.class_weight import compute_class_weight
import numpy as np
import pandas as pd
from tensorflow.keras.layers import Input, Dense, Conv2D, MaxPooling2D, Flatten, Concatenate, Dropout, BatchNormalization
from tensorflow.keras.models import Model
from tensorflow.keras.optimizers import Adam
from sklearn.model_selection import train_test_split
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from sklearn.utils.class_weight import compute_class_weight
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
from tensorflow.keras.utils import Sequence


def build_cancer_net_sca(input_shape=(224, 224, 3), metadata_shape=(5,)):
    # Image input model
    image_input = Input(shape=input_shape)
    conv1 = Conv2D(32, kernel_size=(3, 3), activation='relu')(image_input)
    pool1 = MaxPooling2D(pool_size=(2, 2))(conv1)
    conv2 = Conv2D(64, kernel_size=(3, 3), activation='relu')(pool1)
    pool2 = MaxPooling2D(pool_size=(2, 2))(conv2)
    flatten_img = Flatten()(pool2)

    # Metadata input model
    metadata_input = Input(shape=metadata_shape)
    dense_meta = Dense(32, activation='relu')(metadata_input)

    # Concatenate both branches
    concatenated = Concatenate()([flatten_img, dense_meta])
    x = Dense(64, activation='relu')(concatenated)
    x = BatchNormalization()(x)
    x = Dropout(0.5)(x)
    output = Dense(1, activation='sigmoid')(x)

    model = Model(inputs=[image_input, metadata_input], outputs=output)
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

    return model

# Load your full_metadata
full_metadata['target'] = full_metadata['malignant'].astype(str)  # Convert target to string for compatibility

# Data splitting
train_val_data, test_data = train_test_split(full_metadata, test_size=0.1, random_state=42)
train_data, val_data = train_test_split(train_val_data, test_size=0.1, random_state=42)

# Image Data Generators for augmentation
train_datagen = ImageDataGenerator(
    rescale=1./255,
    rotation_range=20,
    width_shift_range=0.1,
    height_shift_range=0.1,
    shear_range=0.1,
    zoom_range=0.1,
    horizontal_flip=True,
    fill_mode='nearest'
)
val_test_datagen = ImageDataGenerator(rescale=1./255)

# Setup training and validation image generators
train_img_gen = train_datagen.flow_from_dataframe(
    dataframe=train_data,
    x_col='image_path',
    y_col='target',
    target_size=(224, 224),
    batch_size=32,
    class_mode='binary',
    shuffle=True  # Enable shuffling for training data
)
val_img_gen = val_test_datagen.flow_from_dataframe(
    dataframe=val_data,
    x_col='image_path',
    y_col='target',
    target_size=(224, 224),
    batch_size=32,
    class_mode='binary',
    shuffle=False
)

class MultiInputGenerator(Sequence):
    def __init__(self, img_generator, metadata, batch_size):
        self.img_generator = img_generator
        self.metadata = metadata
        self.indices = np.arange(len(metadata))
        self.batch_size = batch_size

    def __len__(self):
        return int(np.ceil(len(self.metadata) / float(self.batch_size)))

    def __getitem__(self, idx):
      inds = self.indices[idx * self.batch_size:(idx + 1) * self.batch_size]
      batch_imgs, labels = self.img_generator.__getitem__(idx)
      batch_meta = self.metadata[inds]  # Adjust indexing here
      return [batch_imgs, batch_meta], labels


    def on_epoch_end(self):
        np.random.shuffle(self.indices)

# Metadata features
metadata_features = ["MEL", "melanocytic", "NV", "benign", "concomitant_biopsy"]
train_metadata = train_data[metadata_features].values
val_metadata = val_data[metadata_features].values

# Build Cancer-Net SCa model
model = build_cancer_net_sca()

# Display model summary
model.summary()

# Prepare training and validation generators
train_gen = MultiInputGenerator(train_img_gen, train_metadata, batch_size=32)
val_gen = MultiInputGenerator(val_img_gen, val_metadata, batch_size=32)

# Define callbacks
early_stopping = EarlyStopping(monitor='val_loss', patience=10, verbose=1, restore_best_weights=True)
model_checkpoint = ModelCheckpoint('best_model.h5', monitor='val_accuracy', save_best_only=True, verbose=1)

# Train the model
history = model.fit(
    train_gen,
    validation_data=val_gen,
    epochs=10,
    callbacks=[early_stopping, model_checkpoint]
)

Found 20517 validated image filenames belonging to 2 classes.
Found 2280 validated image filenames belonging to 2 classes.
Model: "model_8"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input_17 (InputLayer)       [(None, 224, 224, 3)]        0         []                            
                                                                                                  
 conv2d_16 (Conv2D)          (None, 222, 222, 32)         896       ['input_17[0][0]']            
                                                                                                  
 max_pooling2d_16 (MaxPooli  (None, 111, 111, 32)         0         ['conv2d_16[0][0]']           
 ng2D)                                                                                            
                                                                    

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


642/642 [==============================] - 384s 595ms/step - loss: 0.4813 - accuracy: 0.8062 - val_loss: 0.4286 - val_accuracy: 0.8316
Epoch 2/10
642/642 [==============================] - ETA: 0s - loss: 0.4215 - accuracy: 0.8323
Epoch 2: val_accuracy improved from 0.83158 to 0.83728, saving model to best_model.h5
642/642 [==============================] - 383s 596ms/step - loss: 0.4215 - accuracy: 0.8323 - val_loss: 0.4083 - val_accuracy: 0.8373
Epoch 3/10
642/642 [==============================] - ETA: 0s - loss: 0.4023 - accuracy: 0.8385
Epoch 3: val_accuracy did not improve from 0.83728
642/642 [==============================] - 380s 592ms/step - loss: 0.4023 - accuracy: 0.8385 - val_loss: 0.4611 - val_accuracy: 0.8346
Epoch 4/10
642/642 [==============================] - ETA: 0s - loss: 0.4035 - accuracy: 0.8378
Epoch 4: val_accuracy did not improve from 0.83728
642/642 [==============================] - 383s 596ms/step - loss: 0.4035 - accuracy: 0.8378 - val_loss: 0.8068 - val_a

In [128]:
import pandas as pd
from sklearn.model_selection import train_test_split
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.models import load_model
import numpy as np

from tensorflow.keras.utils import Sequence

class MultiInputGenerator(Sequence):
    def __init__(self, img_gen, metadata, batch_size=32):
        self.img_gen = img_gen
        self.metadata = metadata
        self.batch_size = batch_size

    def __len__(self):
        return int(np.ceil(len(self.metadata) / self.batch_size))

    def __getitem__(self, idx):
        batch_x = self.img_gen[idx]  # Get batch of images
        batch_meta = self.metadata.iloc[idx * self.batch_size:(idx + 1) * self.batch_size].to_numpy()
        return [batch_x[0], batch_meta], batch_x[1]  # Return image and metadata inputs together with labels

    def on_epoch_end(self):
        self.img_gen.on_epoch_end()


# Prepare metadata for the test generator, selecting the same indices as images
metadata_columns = ["MEL", "melanocytic", "NV", "benign", "concomitant_biopsy"]  # Example metadata columns used for training
test_metadata = test_data[metadata_columns]  # Extract relevant metadata for test data
test_data['malignant'] = test_data['malignant'].astype(str)

# Setup the original image data generator
test_img_gen = test_datagen.flow_from_dataframe(
    dataframe=test_data,
    x_col='image_path',
    y_col='malignant',
    target_size=(224, 224),
    batch_size=32,
    class_mode='binary',
    shuffle=False
)

# Combine image and metadata into a single generator
test_generator = MultiInputGenerator(test_img_gen, test_metadata)

# Evaluate the model using the combined generator
results = model.evaluate(test_generator)
print(f"Test Loss: {results[0]}, Test Accuracy: {results[1]}")


Found 2534 validated image filenames belonging to 2 classes.
80/80 [==============================] - 22s 272ms/step - loss: 0.5058 - accuracy: 0.8275
Test Loss: 0.5057815313339233, Test Accuracy: 0.8275454044342041


In [129]:
model.save('/content/drive/MyDrive/Colab Models/best_medical_model.keras)'

SyntaxError: incomplete input (<ipython-input-129-bdfcf8d08a95>, line 1)